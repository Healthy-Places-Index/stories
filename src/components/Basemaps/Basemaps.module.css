.closeButton{
  position: absolute;
  right: -10px;
  top: -10px;
  cursor: pointer;
}

.thumbnail{
  width: 50px;
  height: 50px;
  margin-right: 10px;
  background-size: cover;
  background-position: center;
}