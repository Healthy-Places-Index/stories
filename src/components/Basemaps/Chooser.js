import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { Modal, Button, Item } from 'semantic-ui-react';

const Chooser = ({ options, handler }) => {
  const [open, setOpen] = useState(false);

  if (options.length === 0) return <i>No basemaps available for the selected year</i>;

  return (
    <Modal
      closeIcon
      onClose={() => setOpen(false)}
      onOpen={() => setOpen(true)}
      open={open}
      trigger={<Button fluid content="Select Basemap" icon="map outline" labelPosition="left" />}
    >
      <Modal.Header>Select a Basemap</Modal.Header>
      <Modal.Content scrolling>
        <Item.Group divided link>
          {options.map(basemap => (
            <Item
              key={basemap.ssid}
              onClick={() => {
                handler(basemap);
                setOpen(false);
              }}
            >
              <Item.Image size="tiny" src={basemap.thumbnail} />
              <Item.Content>
                <Item.Header style={{ fontSize: 16 }}>{basemap.title}</Item.Header>
                <Item.Description style={{ marginTop: 0 }}>{basemap.creator}</Item.Description>
              </Item.Content>
            </Item>
          ))}
        </Item.Group>
      </Modal.Content>
    </Modal>
  );
};

Chooser.propTypes = {
  options: PropTypes.arrayOf(PropTypes.shape()).isRequired,
  handler: PropTypes.func.isRequired,
};

Chooser.defaultProps = {};

export default Chooser;
