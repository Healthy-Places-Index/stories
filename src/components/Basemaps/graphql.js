import { gql } from '@apollo/client';

export const GET_SLIDE = gql`
  query GetBasemap($slide: ID!) {
    Slide(where: { id: $slide }) {
      id
      year
      opacity
      basemap {
        id
        ssid
        title
      }
    }
  }
`;

export const GET_BASEMAPS = gql`
  query {
    basemaps: allBasemaps {
      id
      ssid
      title
      firstYear
      lastYear
      thumbnail
      creator
    }
  }
`;

export const UPDATE_BASEMAP = gql`
  mutation UpdateBasemap($slide: ID!, $basemap: BasemapRelateToOneInput) {
    updateSlide(id: $slide, data: { basemap: $basemap }) {
      id
      basemap {
        id
      }
    }
  }
`;

export const UPDATE_SLIDE_OPACITY = gql`
  mutation UpdateSlideTitle($slide: ID!, $opacity: Float) {
    updateSlide(id: $slide, data: { opacity: $opacity }) {
      id
      opacity
    }
  }
`;
