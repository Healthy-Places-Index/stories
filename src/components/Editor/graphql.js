import { gql } from '@apollo/client';

export const GET_SLIDES = gql`
  query GetSlide($slide: ID!) {
    Slide(where: { id: $slide }) {
      id
      title
      description
      size
      media
      imageTitle
      source
      url
    }
  }
`;

export const UPDATE_SLIDE_TITLE = gql`
  mutation UpdateSlideTitle($slide: ID!, $title: String) {
    updateSlide(id: $slide, data: { title: $title }) {
      id
      title
    }
  }
`;

export const UPDATE_SLIDE_DESCRIPTION = gql`
  mutation UpdateSlideDescription($slide: ID!, $description: String) {
    updateSlide(id: $slide, data: { description: $description }) {
      id
      description
    }
  }
`;

export const UPDATE_SLIDE_MEDIA = gql`
  mutation UpdateSlideMedia($slide: ID!, $media: String) {
    updateSlide(id: $slide, data: { media: $media }) {
      id
      media
    }
  }
`;

export const UPDATE_SLIDE_SIZE = gql`
  mutation UpdateSlideSize($slide: ID!, $size: SlideSizeType) {
    updateSlide(id: $slide, data: { size: $size }) {
      id
      size
    }
  }
`;

export const ADD_IMAGE = gql`
  mutation AddImage($slide: SlideRelateToOneInput) {
    createImage(data: { slide: $slide }) {
      id
      title
      creator
      source
      date
      url
    }
  }
`;

export const UPDATE_IMAGE = gql`
  mutation UpdateImage($slide: ID!, $imageTitle: String, $source: String, $url: String) {
    updateSlide(id: $slide, data: { imageTitle: $imageTitle, source: $source, url: $url }) {
      id
      imageTitle
      source
      url
    }
  }
`;
