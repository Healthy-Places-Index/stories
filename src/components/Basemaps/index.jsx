import React, { useState, useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import { useQuery, useMutation } from '@apollo/client';
import { Modal, Button, Segment, Item } from 'semantic-ui-react';
import { Slider } from 'react-semantic-ui-range';

import debouncedMutation from '../../lib/debouncedMutation';
import { GET_SLIDE, GET_BASEMAPS, UPDATE_BASEMAP, UPDATE_SLIDE_OPACITY } from './graphql';
import styles from './Basemaps.module.css';

const Basemaps = ({ slide }) => {
  const { data } = useQuery(GET_SLIDE, {
    variables: { slide },
  });
  const allBasemaps = useQuery(GET_BASEMAPS);
  const [updateBasemap] = useMutation(UPDATE_BASEMAP);
  const [updateOpacity] = useMutation(UPDATE_SLIDE_OPACITY);

  const onBasemapChange = newBasemap => {
    let basemap = { disconnectAll: true };
    if (newBasemap) {
      basemap = {
        connect: { id: newBasemap.id },
      };
    }
    updateBasemap({
      variables: {
        slide,
        basemap,
        optimisticResponse: {
          __typename: 'Mutation',
          updateSlide: {
            __typename: 'Slide',
            id: slide,
            basemap: {
              __typename: 'Basemap',
              ...newBasemap,
            },
          },
        },
      },
    });
  };

  const opacityTimer = useRef();

  const [options, setOptions] = useState([]);
  const [opacity, setOpacity] = useState(1);
  const [open, setOpen] = useState(false);

  useEffect(() => {
    if (allBasemaps && allBasemaps.data) {
      const { year } = data.Slide;
      setOptions(allBasemaps.data.basemaps.filter(b => b.firstYear <= year && b.lastYear >= year));
    }
  }, [allBasemaps, data]);

  useEffect(() => {
    opacityTimer.current = debouncedMutation({
      slide,
      timerRef: opacityTimer,
      mutation: updateOpacity,
      values: { opacity },
    });
  }, [opacity]);

  useEffect(() => {
    if (data) {
      setOpacity(data.Slide.opacity);
    }
  }, [allBasemaps.loading, data]);

  return (
    <div>
      <h3 style={{ marginTop: 0 }}>Maps / Plans / Aerials</h3>
      {data && data.Slide.basemap ? (
        <Segment style={{ padding: '0.5em', display: 'flex', alignItems: 'center' }}>
          <div
            className={styles.thumbnail}
            style={{
              backgroundImage: `url(${data.Slide.basemap.thumbnail})`,
            }}
          />
          <div style={{ width: 'calc(100% - 60px)' }}>{data.Slide.basemap.title}</div>
          <Button
            className={styles.closeButton}
            circular
            icon="close"
            size="mini"
            onClick={() => onBasemapChange(null)}
          />
        </Segment>
      ) : (
        <>
          {options.length ? (
            <Modal
              closeIcon
              onClose={() => setOpen(false)}
              onOpen={() => setOpen(true)}
              open={open}
              trigger={
                <Button fluid content="Select Basemap" icon="map outline" labelPosition="left" />
              }
            >
              <Modal.Header>Select a Basemap</Modal.Header>
              <Modal.Content scrolling>
                <Item.Group divided link>
                  {options.map(basemap => (
                    <Item
                      key={basemap.ssid}
                      onClick={() => {
                        onBasemapChange(basemap);
                        setOpen(false);
                      }}
                    >
                      <Item.Image size="tiny" src={basemap.thumbnail} />
                      <Item.Content>
                        <Item.Header style={{ fontSize: 16 }}>{basemap.title}</Item.Header>
                        <Item.Description style={{ marginTop: 0 }}>
                          {basemap.creator}
                        </Item.Description>
                      </Item.Content>
                    </Item>
                  ))}
                </Item.Group>
              </Modal.Content>
            </Modal>
          ) : (
            <i>No basemaps available for the selected year</i>
          )}
        </>
      )}
      <h4>Overlay Opacity</h4>
      <Slider
        disabled={!data || !data.Slide.basemap}
        discrete
        inverted={false}
        value={opacity}
        settings={{
          min: 0,
          max: 1,
          step: 0.1,
          onChange: setOpacity,
        }}
      />
    </div>
  );
};

Basemaps.propTypes = {
  slide: PropTypes.string.isRequired,
};

export default Basemaps;
