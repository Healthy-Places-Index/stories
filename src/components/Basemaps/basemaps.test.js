import React from 'react';
import renderer, { act } from 'react-test-renderer';
import { MockedProvider } from '@apollo/client/testing';
import { GET_BASEMAPS, GET_SLIDE, UPDATE_BASEMAP, UPDATE_SLIDE_OPACITY } from './graphql';

import Basemaps from './index';

const basemap = {
  id: '2',
  ssid: '3',
  title: 'Test',
  thumbnail: 'http://images.imaginerio.org/test.jpg',
};
const mocks = [
  {
    request: {
      query: GET_SLIDE,
      variables: {
        slide: '2',
      },
    },
    result: {
      data: {
        Slide: {
          id: '1',
          year: 1900,
          opacity: 1,
          basemap,
        },
      },
    },
  },
  {
    request: {
      query: GET_BASEMAPS,
    },
    result: {
      data: {
        basemaps: [
          {
            ...basemap,
            firstYear: 1800,
            lastYear: 2000,
          },
        ],
      },
    },
  },
  {
    request: {
      query: UPDATE_BASEMAP,
      variables: { slide: '2', basemap: { connect: { id: '2' } } },
    },
    result: {
      data: {
        id: '2',
        basemap: {},
      },
    },
  },
  {
    request: {
      query: UPDATE_SLIDE_OPACITY,
    },
    result: {
      data: {
        id: '6',
        opacity: 0.5,
      },
    },
  },
];

jest.mock('react-semantic-ui-range', () => {
  const Slider = () => <div className="slider" />;
  return { Slider };
});
jest.mock('semantic-ui-react', () => {
  const Button = props => <div {...props} className="button" />;
  const Segment = props => <div {...props} className="segment" />;
  return { Button, Segment };
});
jest.mock('semantic-ui-react/dist/commonjs/addons/Portal/Portal', () => ({ children }) => children);

describe('Basemaps', () => {
  it('matches snapshot', async () => {
    const component = renderer.create(
      <MockedProvider mocks={mocks} addTypename={false}>
        <Basemaps slide="2" />
      </MockedProvider>
    );

    await act(() => new Promise(resolve => setTimeout(resolve, 0)));
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
