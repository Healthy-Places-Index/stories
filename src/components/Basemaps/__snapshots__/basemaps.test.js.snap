// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basemaps matches snapshot 1`] = `
<div>
  <h3
    style={
      Object {
        "marginTop": 0,
      }
    }
  >
    Maps / Plans / Aerials
  </h3>
  <div
    className="segment"
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "padding": "0.5em",
      }
    }
  >
    <div
      className="thumbnail"
      style={
        Object {
          "backgroundImage": "url(http://images.imaginerio.org/test.jpg)",
        }
      }
    />
    <div
      style={
        Object {
          "width": "calc(100% - 60px)",
        }
      }
    >
      Test
    </div>
    <div
      circular={true}
      className="button"
      icon="close"
      onClick={[Function]}
      size="mini"
    />
  </div>
  <h4>
    Overlay Opacity
  </h4>
  <div
    className="slider"
  />
</div>
`;
